### PR TITLE
Add a safer version of WithNetNS*

### LINKED ISSUE #2419: Fix the usage of WithNetNS* functions
As it's been mentioned in #2388 , the `WithNetNS*` family of functions and their usage are not safe. To address the issues, the work function has to be executed in a separate process which is not associated with the Go runtime process which created it.

Currently, `WithNetNS*` is used by the following functions:

- [x] `common/utils.go:FindNetDevs`
- [x] `common/utils.go:GetWeaveNetDevs`
- [x] `net/veth.go:interfaceExistsInNamespace`
- [x] `net/veth.go:AttachContainer`
- [x] `net/veth.go:DetachContainer`
- [x] `plugin/net/cni.go:CmdAdd`
- [x] `plugin/net/cni.go:CmdDel`

In some cases, shelling out is not necessary - it's enough to move a netns dependant code to the end of execution and to ensure that the code does not create any go-routine which require a particular netns.

### PR DESCRIPTION
**Pull request by brb**
This PR introduces a safe version of `WithNetNS*`. The new function executes `work()` by shelling out `weaveutil` which before executing it enters a given network namespace via `nsenter`.

Also, the PR replaces the usage of `WithNetNS*Unsafe` with the newly introduced function where needed.

Fixes #2419

### CODE REVIEW

**Comment by brb** [comment]
PTAL.

File: net/veth.go [hunk 1]
```diff
@@ -105,21 +105,18 @@ const (
...
-		_, err := netlink.LinkByName(ifName)
-		return err
-	})
+func interfaceExistsInNamespace(nsPath string, ifName string) bool {
+	_, err := WithNetNS(nsPath, "check-iface", ifName)
 	return err == nil
 }
 
-func AttachContainer(ns netns.NsHandle, id, ifName, bridgeName string, mtu int, withMulticastRoute bool, cidrs []*net.IPNet, keepTXOn bool) error {
+func AttachContainer(ns netns.NsHandle, nsPath, id, ifName, bridgeName string, mtu int, withMulticastRoute bool, cidrs []*net.IPNet, keepTXOn bool) error {
```
Comment (rade) [inline net/veth.go]: it's rather odd that this takes both a ns and nsPath.

Comment (rade) [inline net/veth.go]: This also introduces an asymmetry with DetachContainer.

**Comment by bboreham** [comment]
I squashed the commits in this PR, in branch issues/2419-withnetns
If #2541 lands I will rebase on top of that

**Comment by brb** [comment]
PTAL. A few things:

* I'm in favor of getting rid of all occurrences of`WithNetNS*Unsafe` as unaware users might trip over it.
* `ParseNetDev` is ugly. Open to suggestions.
* If moving `NetDev` pieces from `common/` to `net/` clogs reviewing, I will remove the corresponding commit from this PR.

File: prog/weaveutil/netdev.go [hunk 2]
```diff
@@ -33,3 +94,39 @@ func configureARP(args []string) error {
 
 	return nil
 }
+
+// listNetDevs outputs network ifaces identified by the given indexes
+// in the format of weavenet.Dev.
+func listNetDevs(args []string) error {
```
Comment (awh) [inline prog/weaveutil/netdev.go]: I think we should:
* Pass the indices as separate args so you can ditch the `Split`
* Don't bother de-duplicating - the list always comes from `ConnectedToBridgeVethPeerIds` anyway, which can't generate dupes
* Add annotations to the NetDev structure, and print as JSON - that'll fix your ParseNetDev issue at the same time

File: net/netdev.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	Name  string
+	MAC   net.HardwareAddr
+	CIDRs []*net.IPNet
+}
+
+func (d Dev) String() string {
+	return fmt.Sprintf("%s %s %s", d.Name, d.MAC, d.CIDRs)
+}
+
+func ParseNetDev(netdev string) (Dev, error) {
```
Comment (awh) [inline net/netdev.go]: As mentioned elsewhere, this (and the associated regexp/`Dev.String()`) can go away if `list-netdevs` prints JSON

File: net/netdev.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+	peersStr := make([]string, len(peerIDs))
+	for i, id := range peerIDs {
+		peersStr[i] = strconv.Itoa(id)
+	}
+	netdevsStr, err := WithNetNSByPid(processID, "list-netdevs", strings.Join(peersStr, ","))
+	if err != nil {
+		return nil, fmt.Errorf("list-netdevs failed: %s", err)
+	}
+	for _, netdevStr := range strings.Split(netdevsStr, "\n") {
```
Comment (awh) [inline net/netdev.go]: This can be simplified if `list-netdevs` prints JSON encoded `Dev` structures

**Comment by bboreham** [comment]
We don't really need to change `addrs.go` or `proc-addrs.go`; they concentrate all their namespace-changing in one region.  And if you don't do that then a lot of the rest of the complexity goes away.

File: prog/weaveutil/netdev.go (hunk 2 above)
Comment (brb) [inline prog/weaveutil/netdev.go]: Yep, sounds good!

**Comment by brb** [comment]
As mentioned above, I'm in favor of getting rid of `WithNetNS*Unsafe` for safety reasons. Therefore I made the changes to `addrs.go` and `proc-addrs.go`.

**Comment by brb** [comment]
PTAL.

**Pull request merged**